arch/src/x86_64/mod.rs

```diff
@@ -718,6 +723,38 @@ pub fn generate_common_cpuid(
         });
     }
 
+    // mwait
+    // Intel-defined CPU features, CPUID level 0x00000001 (ECX)
+    // SAFETY: leaf are guaranteed to be non-null.
+    #[cfg(feature = "kvm")]
+    if unsafe { std::arch::x86_64::__cpuid(0x1) }.ecx & 1 << X86_FEATURE_MWAIT
+        == 1 << X86_FEATURE_MWAIT
+    {
+        // Thermal and Power Management Leaf, CPUID level 0x00000006 (EAX)
+        // SAFETY: cpuid called with valid leaves
+        if unsafe { x86_64::__cpuid(0x6) }.eax & 1 << X86_FEATURE_ARAT == 1 << X86_FEATURE_ARAT {
+            // Update some existing CPUID for X86_FEATURE_MWAIT
+            for entry in cpuid.as_mut_slice().iter_mut() {
+                if entry.function == 0x1 {
+                    // set mwait bit
+                    if entry.ecx & 1 << X86_FEATURE_MWAIT != 1 << X86_FEATURE_MWAIT {
+                        info!("mwait instruction cpuid leaf enabled.");
+                        entry.ecx |= 1 << X86_FEATURE_MWAIT as u32;
+                    }
+                }
+            }
+
+            // CPUID_MWAIT_LEAF
+            let function = 0x5;
+            // SAFETY: leaf are guaranteed to be non-null.
```

Comment:
Does it mean that CPUID leaf 0x5 won't be null as long as the host CPU advises the feature `X86_FEATURE_MWAIT` (via CPUID LEAF 0x01 ECX)?
Also, I think it reads better (and is consistent with the rest of the code if you do: `let leaf = unsafe { std::arch::x86_64::__cpuid(0x5) };`

Reply:
Yes, if we want to forward the mwait feature to the guest, we must make sure that CPUID leaf 0x5 won't be null. Otherwise, the initialization of intel cpu idle driver in the guest kernel will fail.